pmd-javascript/src/main/resources/category/ecmascript/performance.xml

```diff
@@ -9,4 +8,38 @@
 Rules that flag suboptimal code.
     </description>
 
+    <rule name="AvoidConsoleStatements"
+          language="ecmascript"
+          since="7.4.0"
+          message="Avoid console statements since they negatively impact performance"
+          class="net.sourceforge.pmd.lang.rule.xpath.XPathRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_ecmascript_performance.html#avoidconsolestatements">
+        <description>
+Using the console for logging in production might negatively impact performance.
+In addition, logging could expose sensitive data.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="methods" type="List[String]" value="log,error,info,warn,debug,trace" description="The methods of the console object that should be flagged."/>
```

Comment:
I'd also include `dir`, `dirxml`, `table`, `profile` (non-standard)

Reply:
If we add these non-standard methods, then we should add a lot more:
https://developer.mozilla.org/en-US/docs/Web/API/console

Are there any methods, that should be allowed to use or should the rule actually ban the usage of console at all?
@dschach What exactly did you have in mind, when you requested #5105?

Reply:
For some systems (for Salesforce lightning web components in my situation), it is possible for a developer to expose database records in a browser console log. This can be a security violation for, as an example, medical records.

At a minimum, the capability to prevent any call to the console object would be a good screen for this.

As for the specifics, that's an interesting feature that I hadn't thought of. 

It seems that the [console.log()](https://developer.mozilla.org/en-US/docs/Web/API/console/log_static), [console.info()](https://developer.mozilla.org/en-US/docs/Web/API/console/info_static), [console.warn()](https://developer.mozilla.org/en-US/docs/Web/API/console/warn_static), [console.error()](https://developer.mozilla.org/en-US/docs/Web/API/console/error_static), and [console.debug()](https://developer.mozilla.org/en-US/docs/Web/API/console/debug_static) methods are the ones that could take data and expose it. The others just format it, if I'm reading correctly.

Reply:
The following methods could potentially expose data:
- assert() - you can provide a "msg" parameter
- count() - you can provide a label that will be printed in the console
- debug()
- dir()
- dirxml() - seems to behave the same as dir() at least in Firefox
- error()
- exception() - non-standard, same as error()
- group() - you can provide a label that will be printed in the console
- groupCollapsed() - same as group()
- info()
- log()
- table()
- timeEnd() - given label will be printed in the console
- timeLog() - given label will be printed in the console
- trace()
- warn()

The following methods impact performance only:
- profile() - non-standard, could indicate left-overs in the code from a profiling session
- time()
- timeStamp() - non-standard

The following methods seem to be "safe" regarding information disclosure:
- clear()
- countReset()
- groupEnd()

So, almost all methods are in some way or the other prone to performance or information disclosure. Maybe it's indeed better to just flag all usages of console and see how it goes. We can always improve the rule later and add a property. But removing a property later would be an incompatible change.

Reply:
I've update the rule now, so that it flags all usages of console, regardless of the method.

In terms of exposing sensitive data: if you can output it via console.log, then it is already exposed... it might be easier for an attacker, if some data is directly visible, but: not using console.log doesn't avoid exposing data - if send too much data from the backend to the frontend, then your backend is exposing sensitive data... So, I think the category "performance" is indeed better than "security".